### PR TITLE
New process for handling virtualization proposals

### PR DESCRIPTION
**Pull request by vladikr**
**What this PR does / why we need it**:

This PR is based on #288 and similarly suggests creating a new `kubevirt/enhancements `repository to manage KubeVirt Enhancement Proposals (VEPs). 
However, this proposal suggests a structured process for prioritizing enhancements, focusing review and development efforts, and improving collaboration across Special Interest Groups (SIGs). 
Each VEP will serve as the single source of truth for its feature, tracking its design, progress, and associated bugs, ensuring transparency and focus during development cycles.

**Release note**:

```release-note
None
```

### CODE REVIEW

File: design-proposals/new-proposals-framework.md
Comment (dankenigsberg) [inline design-proposals/new-proposals-framework.md:37]: Can we add here a link to a VEP template, or a reference to KEP's?

File: design-proposals/new-proposals-framework.md
Comment (dankenigsberg) [inline design-proposals/new-proposals-framework.md:40]: Can we add a template for such issue?

Comment (dankenigsberg) [inline design-proposals/new-proposals-framework.md:40]:
```suggestion
4. **Visibility and Tracking**: The Author of an accepted VEPs will open an issue to track their progress, maturity stages (alpha, beta, GA), list the associated bugs and user feedback
```

File: design-proposals/new-proposals-framework.md
Comment (dankenigsberg) [inline design-proposals/new-proposals-framework.md:42]:
```suggestion
5. *Single source of truth**: Each VEP will be the authoritative reference for the associated feature. This aligns with the Kubernetes KEP process. It will ensure that each enhancement
includes all the relevant information, including the design and the state.
The VEP owner is responsible to update it as its development progresses, until it is fully mature (or deprecated). 
```

File: design-proposals/new-proposals-framework.md
Comment (dankenigsberg) [inline design-proposals/new-proposals-framework.md:55]:
```suggestion
- Luboslav Pivarc @xpivarc 
- Vladik Romanovsky @vladikr 
```

**Review by dankenigsberg — APPROVED** [review]

File: design-proposals/new-proposals-framework.md
Comment (fabiand) [inline design-proposals/new-proposals-framework.md:53]: How did we end up with this list?
What is the reasoning and the tradeoffs?

**Review by fabiand — COMMENTED** [review]
Great to see this topic being pushed forward again.

File: design-proposals/new-proposals-framework.md
Comment (vladikr) [reply design-proposals/new-proposals-framework.md:39]: Effectively, all the approved proposals before the dev cycle will become the projects' priority for that cycle. PRs belonging to the approved proposals will be prioritized (dev and reviews bandwidth) 
If there are too many, we will prioritize them during the unconference. sigs will set the priority.

File: design-proposals/new-proposals-framework.md
Comment (vladikr) [reply design-proposals/new-proposals-framework.md:53]: This is only a suggested list.

As I mentioned in the document, the goal is to gradually transfer responsibility to the sigs as they mature and show commitment to the project as a whole, not only to the specific group.

To kickstart this process, we need a small group of people who will be committed to this process and who are not acting as SIG chairs.
I don't see a benefit in having a large list of approvers who are effectively inactive.

If anyone else would like to commit to this work, they are more than welcome to step up and add a name here.

File: design-proposals/new-proposals-framework.md
Comment (iholder101) [reply design-proposals/new-proposals-framework.md:39]: +1 for SIGs being in charge of the prioritization.

A small note: when I say prioritization I think mostly on reviews, and not on dev work. The dev work should be done by whoever is interested in implementing the proposal, but the review must be done by sig/root approvers. In other words, as Kubevirt maintainers, we owe VEP reviews, but we are not obligated to implement them.

File: design-proposals/new-proposals-framework.md
Comment (iholder101) [inline design-proposals/new-proposals-framework.md:42]: Maybe we can emphasize here that whenever the design / api / limitations etc change the VEP should be updated as well with a PR

File: design-proposals/new-proposals-framework.md
Comment (iholder101) [reply design-proposals/new-proposals-framework.md:53]: I'm also conflicted to whether we need a small list of approvers for this work instead of simply all of the sig approvers. Anyhow each VEP will be owned by a target SIG, as mentioned above.

I'd at least add here that we aim that in the future this work would be done by SIGs, which I see as part of their core responsibility.

File: design-proposals/new-proposals-framework.md
Comment (vladikr) [reply design-proposals/new-proposals-framework.md:40]: Eventually, we will. I don't think we should overload this PR with this. 
I think eventually it should look similar to the k8s tracker: https://github.com/kubernetes/enhancements/issues/3521

File: design-proposals/new-proposals-framework.md
Comment (vladikr) [reply design-proposals/new-proposals-framework.md:54]: The goal is to eventually shift responsibilities toward the SIGs. However, this PR not only proposes changing the mechanics of how we manage proposals in KubeVirt but also asks the SIGs to adopt a different mindset.

As SIGs take on more control over VEPs, it is essential for the approvers to prioritize the benefit of the entire project over their downstream priorities or the interests of individual SIGs.
Unfortunately, there are many examples where this was not the case.

This is why I think we need a transition period that would allow SIGs to mature and adjust while keeping the centralized oversight.

FWIW, we could add here the current list of the community repo approvers, excluding those who are conflicted.

File: design-proposals/new-proposals-framework.md
Comment (alicefr) [reply design-proposals/new-proposals-framework.md:42]: Who is responsible to keep the VEP in sync with the current state of the code?

Comment (iholder101) [reply design-proposals/new-proposals-framework.md:42]: > Who is responsible to keep the VEP in sync with the current state of the code?

As I see it, and this is how it works in Kubernetes as well, the workflow should look something like this:
1. The developer / assignee proposes a VEP, which contains requirements for graduating from alpha to beta.
2. The VEP is approved and merged.
3. The assignee is responsible for implementation.
4. Once the feature is ready to graduate to Beta from the assignee's perspective, a PR will be introduced to update the VEP. The PR will set a target release for Beta graduation and will explain how the requirements for Beta are fulfilled and whether they have slightly changed. The assignee will possibly have to prove it by providing a list of major implementation PRs.
5. Approvers would need to review the VEP, become convinced that the requirements are both good enough for graduation and are truly fulfilled.
6. The VEP will be updated, the feature would graduate.

TLDR: There are two parties here. One is the assignee who's the one having the biggest interest / motivation to get the feature towards graduation. The other is the SIG/project maintainers who focus on graduating only mature enough features.

Therefore, It's the assignee's responsibility to make the updates, and maintainer's responsibility to review and ensure it is properly done.

File: design-proposals/new-proposals-framework.md
Comment (iholder101) [reply design-proposals/new-proposals-framework.md:40]: We do have this: https://github.com/kubevirt/community/blob/main/design-proposals/feature-lifecycle.md#releases.

Some of the k8s characteristics I remember from the top of my head:
1. Beta features are (usually) on by default.
2. Alpha can be dropped without a warning and has no backward compatibility considerations whatsoever.
3. When there's a good reason for it, there could be sub-stages like Beta1, Beta2, etc.
4. The changes from BetaN (N being the latest beta) to GA need to be minimal, especially API changes.

File: design-proposals/new-proposals-framework.md
Comment (vladikr) [reply design-proposals/new-proposals-framework.md]: @lyarwood has already created a proposal for a template: https://github.com/kubevirt/kubevirt/pull/13167
I think, we need to move it to the new project, once it's created.

File: design-proposals/new-proposals-framework.md
Comment (alaypatel07) [reply design-proposals/new-proposals-framework.md:42]: FWIW, it could be possible that assignees change during the lifecycle of the feature (even k8s suffers from this issue). In which case the burden will fall on to SIGs (esp members with admin rights to the repository) to update the issue and keep it in sync.

Comment (iholder101) [reply design-proposals/new-proposals-framework.md:42]: @alaypatel07 I agree.
It's important to emphasize that if the assignee at some point is not willing to promote the VEP anymore, it's the SIG's responsibility to reflect that and keep the VEP in sync, but it's not its responsibility to continue the implementation. If there is enough interest and another assignee is willing to step in, that's great, but if not, the SIG can eventually decide to revert and drop the VEP.

File: design-proposals/new-proposals-framework.md
Comment (lyarwood) [reply design-proposals/new-proposals-framework.md:39]: > Effectively, all the approved proposals before the dev cycle will become the projects' priority for that cycle. PRs belonging to the approved proposals will be prioritized (dev and reviews bandwidth) If there are too many, we will prioritize them during the unconference. sigs will set the priority.

The dev cycle opens once we hit feature freeze and branch main for the previous release. As such we need to be careful suggesting that only approved proposals will be prioritised by the unconference as that might not be enough time for folks to pivot from the previous release.

IMHO the unconference is just an opportunity to discuss and refine the list of possible proposals for the release. I think I've suggested previously that proposals being fully approved should instead be a goal by a later milestone such as alpha. Using v1.5.0 as an example that could look something like this:

- 2024-10-22 v1.4.0 FF (VEPs for v1.5.0 can be drafted before or after this when folks have bandwidth)
- 2024-10-29 v1.5.0 Unconference (VEP discussions, refinement and initial prioritisation)
- 2024-12-18 v1.5.0 Alpha (VEP approval deadline)
- 2025-01-15 v1.5.0 Beta0
- 2025-02-12 v1.5.0 FF

There's definitely an argument to have the unconference a few weeks later so folks have more time to prepare VEPs but we would still want time after the unconference to let any unresolved issues be worked out before the alpha deadline.

File: design-proposals/new-proposals-framework.md
Comment (lyarwood) [inline design-proposals/new-proposals-framework.md:60]: The dev cycle for `v1.5.0` is already underway so just merge this with the second bullet?

File: design-proposals/new-proposals-framework.md
Comment (vladikr) [reply design-proposals/new-proposals-framework.md:39]: Thank you @lyarwood !
I fully agree with this. We need to start and adjust as we go.

**Review by lyarwood — APPROVED** [review]

**Review by enp0s3 — APPROVED** [review]
@vladikr Thank you! going over the process and the comments, it makes sense to me.

File: design-proposals/new-proposals-framework.md
Comment (dankenigsberg) [reply design-proposals/new-proposals-framework.md:39]: +1 on elaboration of the term "mechanism". I'd say that the VEP proposers and approvers would publicly discuss the relative priority of the VEPs and decide which of them are scheduled to the upcoming release.

Comment (vladikr) [reply design-proposals/new-proposals-framework.md:39]: I merely wanted to say that all accepted VEPs will become the projects' priority. 

The acceptance will be done via a central function. 
If there is support for the VEP in the community and someone who is committed to implementing it, it should be accepted.
PRs belonging to these accepted VEPs will be prioritized. Meaning that there shouldn't be a relative priority between the accepted VEPs. 

Let me try to rework this bullet point.

Comment (enp0s3) [reply design-proposals/new-proposals-framework.md:39]: I noticed that in k8s there are two types of documents:
1. [Enhancement tracking and backlog](https://github.com/kubernetes/enhancements/blob/master/README.md)
2. [How to target a contribution (EP, bugfix, etc.) into a release milestone.](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-release/release.md)

IIUC this design proposal is similar to 1 while the centralized prioritization is also related to our release process. 
Maybe we should create a separate doc that describes our release procedure and how VEP should interact with it.

File: design-proposals/new-proposals-framework.md
Comment (Barakmor1) [inline design-proposals/new-proposals-framework.md:7]: +1

File: design-proposals/new-proposals-framework.md
Comment (vladikr) [reply design-proposals/new-proposals-framework.md]: done

File: design-proposals/new-proposals-framework.md
Comment (vladikr) [reply design-proposals/new-proposals-framework.md]: @dankenigsberg I reworded this.

Comment (vladikr) [reply design-proposals/new-proposals-framework.md]: @enp0s3 we will definitely need to update an existing doc or write a new one. I'm not sure yet. Maybe @aburdenthehand will advice once this PR is merged.

**Comment by vladikr** [comment]
I believe I addressed all existing comments. 
I'd prefer not to approve my own PR (perhaps I had to let someone else post this).
If we all agree to start with his process let's approve this before the v1.5 alpha deadline.

**Review by enp0s3 — APPROVED** [review]

File: design-proposals/new-proposals-framework.md
Comment (Barakmor1) [inline design-proposals/new-proposals-framework.md:27]: Out of curiosity, why not remove the approvals from the community repository right away? Maybe we could consider making the transition in phases, though I'm not sure exactly how that would work.

**Review by Barakmor1 — COMMENTED** [review]
In general, I believe we need to update the current process, and this seems like a step in the right direction. Thank you!

File: design-proposals/new-proposals-framework.md
Comment (fossedihelm) [inline design-proposals/new-proposals-framework.md:40]: Which repo will these issues be filled? `kubevirt/enhancements`?

File: design-proposals/new-proposals-framework.md
Comment (fossedihelm) [inline design-proposals/new-proposals-framework.md:42]:
```suggestion
5. **Single source of truth**: Each VEP will be the authoritative reference for the associated feature. This aligns with the Kubernetes KEP process. It will ensure that each enhancement includes all the relevant information, including the design and the state.
```

File: design-proposals/new-proposals-framework.md
Comment (fossedihelm) [inline design-proposals/new-proposals-framework.md:43]: As mentioned above by others, I think we need to emphasize that, in case of the author's unresponsiveness, the fallback will be the sig itself.

**Review by fossedihelm — COMMENTED** [review]
Looks great overall! I agree with the proposed process. Few comments below about details, otherwise LGTM

**Comment by iholder101** [comment]
Thanks a lot @vladikr. This is a very important step in the right direction.

/lgtm

**Review by dankenigsberg — APPROVED** [review]

**Review by lyarwood — APPROVED** [review]

**Comment by xpivarc** [comment]
+1

File: design-proposals/new-proposals-framework.md
Comment (vladikr) [reply design-proposals/new-proposals-framework.md:40]: Yeah, exactly! 
I commented about this in the thread. Lee has already a PR ready with the new template that we will just move to the new repo.

File: design-proposals/new-proposals-framework.md
Comment (vladikr) [reply design-proposals/new-proposals-framework.md]: Done

**Comment by fossedihelm** [comment]
Thank you!
/lgtm

**Review by jean-edouard — COMMENTED** [review]
/approve

File: design-proposals/new-proposals-framework.md
Comment (fossedihelm) [reply design-proposals/new-proposals-framework.md:43]: Agree! Thank you!

**Pull request merged**
